File: src/main/java/random/telegramhomebot/auth/controllers/RestResponseEntityExceptionHandler.java
package random.telegramhomebot.auth.controllers;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindException;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
import random.telegramhomebot.auth.dao.GenericResponse;
import random.telegramhomebot.auth.exceptinos.InvalidOldPasswordException;
import random.telegramhomebot.services.messages.MessageService;

@RequiredArgsConstructor
@ControllerAdvice
public class RestResponseEntityExceptionHandler extends ResponseEntityExceptionHandler {

    private final MessageService messageService;

    // 400
    @Override
    protected ResponseEntity<Object> handleBindException(
            final BindException ex, final HttpHeaders headers, final HttpStatus status, final WebRequest request) {
        logger.error("400 Status Code", ex);
        final BindingResult result = ex.getBindingResult();
        final GenericResponse bodyOfResponse = new GenericResponse(result.getAllErrors(), "Invalid" + result.getObjectName());
        return handleExceptionInternal(ex, bodyOfResponse, new HttpHeaders(), HttpStatus.BAD_REQUEST, request);
    }

    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(
            final MethodArgumentNotValidException ex,
            final HttpHeaders headers, final HttpStatus status, final WebRequest request) {
        logger.error("400 Status Code", ex);
        final BindingResult result = ex.getBindingResult();
        final GenericResponse bodyOfResponse = new GenericResponse(result.getAllErrors(), "Invalid" + result.getObjectName());
        return handleExceptionInternal(ex, bodyOfResponse, new HttpHeaders(), HttpStatus.BAD_REQUEST, request);
    }

    @ExceptionHandler({InvalidOldPasswordException.class})
    public ResponseEntity<Object> handleInvalidOldPassword(final RuntimeException ex, final WebRequest request) {
        logger.error("400 Status Code", ex);
        final GenericResponse bodyOfResponse
                = new GenericResponse(messageService.getMessage("message.invalidOldPassword"), "InvalidOldPassword");
        return handleExceptionInternal(ex, bodyOfResponse, new HttpHeaders(), HttpStatus.BAD_REQUEST, request);
    }

    @ExceptionHandler({Exception.class})
    public ResponseEntity<Object> handleInternal(final RuntimeException ex, final WebRequest request) {
        logger.error("500 Status Code", ex);
        final GenericResponse bodyOfResponse
                = new GenericResponse(messageService.getMessage("message.error"), "InternalError");
        return new ResponseEntity<>(bodyOfResponse, new HttpHeaders(), HttpStatus.INTERNAL_SERVER_ERROR);
    }
}

File: src/main/java/random/telegramhomebot/auth/controllers/RestResponseEntityExceptionHandler.kt
package random.telegramhomebot.auth.controllers

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.context.request.WebRequest
import random.telegramhomebot.auth.dto.GenericResponse
import random.telegramhomebot.auth.exceptinos.InvalidOldPasswordException
import random.telegramhomebot.services.messages.MessageService

@ControllerAdvice
class RestResponseEntityExceptionHandler(private val messageService: MessageService) {

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValid(ex: MethodArgumentNotValidException, request: WebRequest): ResponseEntity<Any> {
        val result = ex.bindingResult
        val bodyOfResponse = GenericResponse(result.allErrors, "Invalid: ${result.objectName}")
        return ResponseEntity(bodyOfResponse, HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(InvalidOldPasswordException::class)
    fun handleInvalidOldPassword(ex: InvalidOldPasswordException, request: WebRequest): ResponseEntity<Any> {
        val bodyOfResponse =
            GenericResponse(messageService.getMessage("message.invalidOldPassword"), "InvalidOldPassword")
        return ResponseEntity(bodyOfResponse, HttpStatus.BAD_REQUEST)
    }

    @ExceptionHandler(Exception::class)
    fun handleAllExceptions(ex: Exception, request: WebRequest): ResponseEntity<Any> {
        val bodyOfResponse =
            GenericResponse(messageService.getMessage("message.error"), "Internal Server Error")
        return ResponseEntity(bodyOfResponse, HttpStatus.INTERNAL_SERVER_ERROR)
    }
}